docs/jmespath-grammar.txt

```diff
@@ -0,0 +1,135 @@
+eq              = ws "==" ws
+ne              = ws "!=" ws
+expref-token    = "&" ws
+not-token       = "!" ws
+begin-filter    = "[?" ws
+flatten         = ws "[]" ws
+
+; Insignificant whitespace
+ws = *(%x20 / ; Space
+       %x09 / ; Horizontal tab
+       %x0A / ; Line feed or New line
+       %x0D) ; Carriage return
+
+; "&&" binds more tightly than "||"
+; "||" binds more tightly than "|".
+terminal     = pipe / or / and
+non-terminal = root-expr / wildcard-values
+pipe         = expr pipe-separator ( non-terminal / or / and )
+or           = ( non-terminal / and ) or-separator ( non-terminal / or / and )
+and          = non-terminal and-separator ( non-terminal / and )
+
+; subexpr handles most that descend into nodes.
+subexpr            = object-subexpr / array-subexpr
+object-subexpr     = object-subexpr-lhs decimal-point object-subexpr-rhs
+object-subexpr-lhs = subexpr / index / flatten / wildcard-index /
+                     filter / identifier / current-node / literal /
+                     multi-hash / function / group / slice / wildcard-values
+object-subexpr-rhs = identifier / multi-list / multi-hash / function /
+                     wildcard-values
+array-subexpr      = array-subexpr-lhs array-subexpr-rhs
+array-subexpr-lhs  = subexpr / index / flatten / wildcard-index / filter /
+                     identifier / current-node / literal / root-multi-list /
+                     function / group / slice / wildcard-values / literal
+array-subexpr-rhs  = index / slice / wildcard-index / flatten / filter
+
+; Array related rules
+index           = begin-array number end-array
+slice           = begin-array [number] name-separator
+                               [number] [name-separator [number] ]
+                               end-array
+wildcard-index   = begin-array "*" end-array
+filter           = begin-filter filter-condition end-array
+filter-condition = not / non-test / terminal / comparison
+
+not             = not-token ( non-test / not )
+comparison      = non-terminal comparator ( non-test / not / wildcard-values )
+comparator      = lt / lte / gt / gte / eq / ne
+
+root-multi-list = begin-array
+                  ( root-expr / multiple-values / terminal )
+                  end-array
+multi-list      = begin-array ( expr / multiple-values ) end-array
+multiple-values = expr 1*( value-separator expr )
+
+multi-hash      = begin-object ( keyval *( value-separator keyval ) ) end-object
+keyval          = identifier name-separator expr
+
+function        = unquoted-string arg-list
+arg-list        = begin-group [ arg *( value-separator arg ) ] end-group
+arg             = expr / expref
+
+expref          = expref-token expr
+wildcard-values = "*"
+current-node    = "@"
+
+number            = ["-"] int
+
+; Strings, identifiers, and characters.
+raw-string        = "'" *raw-string-char "'"
+raw-string-char   = (%x20-26 / %x28-5B / %x5D-10FFFF) / raw-string-escape
+raw-string-escape = escape ["'"]
```

Comment:
Just a heads up, I synced with this latest version that includes raw string literals and abnfstress caught this.  The `["'"]` means that the closing single quote is optional so strings such as `'\'` would be valid.  This should be ``raw-string-escape = escape "'"``